### ISSUE TITLE
Proffer with Angular

### ISSUE THREAD

**Issue by spacebiscuit** [davidyell/CakePHP3-Proffer#273]
I am using Angular on my front end and Cake's CRUD plugin to provide an API on the backend. When I try to create a new entity with related images I get this error:

`Notice (8): Undefined index: tmp_name [APP/Model\Validation\ProfferRulesTest.php, line 26]`

The image object i amposting has the following structure:

```
lastModified: 1490387993409
lastModifiedDate: Fri Mar 24 2017 20:39:53 GMT+0000 (Greenwich Mean Time) {}
name: "20170324_185515.jpg"
size: 155232
type: "image/jpeg"
webkitRelativePath: ""
```

I believe this is webkit and tmp_name is a php file standard. Is it possible to make the two work together. I could massage the object before I pass it to the backend perhaps?

**Comment by davidyell** [comment]
Thanks for submitting a ticket. I had not considered this use-case before, having not used much JS as a front-end. It is focused, as you rightly say, on being posted from PHP.

I would think that updating the request data before it is marshalled would be a good solution.

**Comment by spacebiscuit** [comment]
I'll give it a go - thanks for tip on where to start. I'll post my findings.

**Comment by spacebiscuit** [comment]
I'm not sure how I can set the tmp_name manually with a valid value because although I can easily set an arbitrary value how can I know the random name used for each image temporarily stored, for example these are town examples:

'tmp_name' => 'C:\wamp64\tmp\php6B85.tmp',
'tmp_name' => 'C:\wamp64\tmp\php7318.tmp',

If this is set by PHP how can it be set manually in my Cake app?

**Comment by davidyell** [comment]
Hmm, perhaps it's worth looking at https://github.com/FriendsOfCake/cakephp-upload and using a custom Writer class and a custom transformer to generate thumbnails using a library of your choice.

**Comment by spacebiscuit** [comment]
So actually this is a lot simpler than I imagined and all the work which is required is on the front-end side. 

What I had to do was use a FormData object to set the file data which then resulted in the file attributes are passed before make the call to the Cake backend:

```
var dataObj = new FormData();

angular.forEach($scope.listing.listing_images, function(value, key) {
     dataObj.append('listing_images[' + key + ']', value); 
});

$http({
     method: 'post',
     url: frontbaseurl + '/api/listings.json',
     data: dataObj,
     headers: {'Content-Type': undefined},
});
```

This wasn't obvious to me because to date when posting to the back-end I have simply been passing an object with attributes for each field to be saved. The idea of using a formData object is new but seems to be the solution.

Feel free to close this 'issue'. Thanks!
